### ISSUE TITLE
Wrong results from field when forcedType references both converter and binding

### ISSUE THREAD

**Issue by adrianmsmith** [jOOQ/jOOQ#8241]
### Short description

Create a field using both a Converter together with a Binding. Do a SELECT. Instead of reading the value of the field (which uses both the Converter and the Binding), you see the value of the previous field. So "select a,b" and read value2() should give you b but instead it gives you a.

### Steps to reproduce the problem

See MCVE at https://github.com/adrianmsmith/jOOQ-binding-converter-bug

### Short description of the setup of the test

Database:
```
CREATE TABLE test (id INT,  my_column VARCHAR(100));
INSERT INTO test VALUES (42, 'StringInDatabase');
```
jooq config:
```
<forcedType>
    <userType>java.lang.String</userType>
    <binding>foo.PrefixFooBinding</binding>
    <converter>foo.PrefixBarConverter</converter>
    <expression>MY_COLUMN</expression>
</forcedType>
```
The "PrefixFooBinding" takes the database String and adds "Foo" before it. The "PrefixBarBinding" takes the String and adds "Bar" before it. Perhaps not the most realistic of examples but I think it serves to demonstrate the issue.

Test code:
```
Record2<Integer, String> x = ctx.select(TEST.ID, TEST.MY_COLUMN).from(TEST).fetchOne();
assertEquals("BarFooStringInDatabase", x.value2());
```

### Expected behavior

Test passes.

### Actual behavior

```
Failed tests:   mcveTest(org.jooq.mcve.test.MCVETest): 
expected:<BarFooStringInDatabase> but was:<42>
````

### Analysis

As far as I can see the code in [DefaultBinding](https://github.com/jOOQ/jOOQ/blob/bb836378ba57c36c6c2c9cfd19d9d9dd2f9a067f/jOOQ/src/main/java/org/jooq/impl/DefaultBinding.java#L360) takes a `ctx` then uses `ctx.convert(..)` to create a converter which references it. When the `binding.set` method gets called, the converter (in normal examples) uses a line like:

        ctx.convert(converter()).value(ctx.resultSet().getString(ctx.index()));

So this calls *another* `convert(..)` method; I think the expectation is that you can chain `convert(..)` as long as you need to, but in fact you can only chain them to a maximum level of 1. That means that using just a Binding or just a Converter (or neither) works, but using both together does not.

The `convert` [method](https://github.com/jOOQ/jOOQ/blob/d75ac1a799140ff9f554923a828f52642ad9408c/jOOQ/src/main/java/org/jooq/impl/DefaultBindingGetResultSetContext.java#L87) on a Converter creates an inner class which writes into its (outer class) value. So calling A.convert will yield B which sets A's value. But using B.convert yields C which writes into B's value, so A's value is left unchanged.

The [CursorImpl](https://github.com/jOOQ/jOOQ/blob/3f9cd17d34bef0d5c23cfba19a8a166dd8c0417a/jOOQ/src/main/java/org/jooq/impl/CursorImpl.java#L1708) creates an object to hold column index and value, goes through the fields, and for each field updates the column index and writes the value. The field column index does get updated, but the value never gets overwritten from the previous column, meaning that the value of the first column (42) is presented as the value of the second column.

### Is this even supposed to work?

The documentation covers Converters and Bindings, but (as far as I can see) neither explicitly states that Converters may be used together with Bindings, nor explicitly states that they can't be.

However the source code for DefaultBinding seems to have [explicit code](https://github.com/jOOQ/jOOQ/blob/bb836378ba57c36c6c2c9cfd19d9d9dd2f9a067f/jOOQ/src/main/java/org/jooq/impl/DefaultBinding.java#L334) for handling the case that a binding is used together with a converter, so I'm assuming it is supposed to be supported.

### Versions:

- jOOQ: 3.11.7 (also tested on 3.10.x)
- Java: 1.8
- Database (include vendor): Any (MVCE uses h2, my original problem used MySQL)
- OS: macOS

### Thanks

Off-topic: I have been using jOOQ full-time on multiple projects for my customers for the last 3-4 years. Alas only using MySQL and PostgreSQL so I wasn't able to persuade my customers to purchase a license yet (but always looking out for a way to!) This is the first time I've ever found a bug or something that didn't work as expected. Otherwise jOOQ is extremely helpful and has always "just worked". On one project recently I access a customer's DB and they changed their schema quite radically - regenerate and recompile and see all the errors, brilliant! You have my gratitude!

### Fix

I do not have a fix (I have a workaround for my particular application), but I would be happy to try to create one and create a PR if you're interested.

**Comment by lukaseder** [comment]
Thanks a lot for your report and for providing the MCVE. I can reproduce this issue. Will investigate.

> The documentation covers Converters and Bindings, but (as far as I can see) neither explicitly states that Converters may be used together with Bindings, nor explicitly states that they can't be.

It is an edge case, which was deliberately not documented, as there are some open questions regarding the order of applications of converters and bindings. It may well be that the resolution here is: Behaviour is undefined. But I'll have to review the existing logic first, to confirm.

> However the source code for DefaultBinding seems to have explicit code for handling the case that a binding is used together with a converter, so I'm assuming it is supposed to be supported.

There are some internal cases where this may make sense, but not necessarily for user code.

> Off-topic: I have been using jOOQ full-time on multiple projects for my customers for the last 3-4 years. Alas only using MySQL and PostgreSQL so I wasn't able to persuade my customers to purchase a license yet (but always looking out for a way to!)

It's alright, that's how we work - good for you and your customers :) And I'm sure in a future project, you'll be using Oracle or SQL Server, so don't worry...

> This is the first time I've ever found a bug or something that didn't work as expected. Otherwise jOOQ is extremely helpful and has always "just worked". On one project recently I access a customer's DB and they changed their schema quite radically - regenerate and recompile and see all the errors, brilliant! You have my gratitude!

That's really great to hear, thank you for the nice words!

**Comment by lukaseder** [comment]
This patch seems to fix it:
[8241.patch.txt](https://github.com/jOOQ/jOOQ/files/2802582/8241.patch.txt)

The problem is that `DefaultBindingGetResultSetContext` (and other similar types) set the value on their own reference, instead of the original one, which is used from `CursorImpl`. This doesn't appear to be an issue if only either converter or binding are used.

**Comment by lukaseder** [comment]
Fixed for jOOQ 3.12. Backports scheduled for 3.11.10 (#8258) and 3.10.9 (#8259)
